File: EntreNosApp/src/ProjetoEncontros.AplicativoWeb/lib/funcionalidades/combinados/modelos/item_do_encontro.dart
class ItemDoEncontro {
  const ItemDoEncontro({
    required this.identificador,
    required this.identificadorDoEncontro,
    required this.descricao,
    required this.situacao,
    required this.identificadorDoUsuarioQueCriou,
    required this.usuarioAtualEhResponsavel,
    required this.criadoEm,
    required this.atualizadoEm,
    this.identificadorDoUsuarioResponsavel,
    this.nomeDoResponsavel,
    this.urlDaFotoDePerfilDoResponsavel,
  });

  factory ItemDoEncontro.deJson(Map<String, dynamic> json) {
    return ItemDoEncontro(
      identificador: json['identificador'] as String,
      identificadorDoEncontro: json['identificadorDoEncontro'] as String,
      descricao: json['descricao'] as String,
      situacao: json['situacao'] as String,
      identificadorDoUsuarioQueCriou:
          json['identificadorDoUsuarioQueCriou'] as String,
      identificadorDoUsuarioResponsavel:
          json['identificadorDoUsuarioResponsavel'] as String?,
      nomeDoResponsavel: json['nomeDoResponsavel'] as String?,
      urlDaFotoDePerfilDoResponsavel:
          json['urlDaFotoDePerfilDoResponsavel'] as String?,
      usuarioAtualEhResponsavel:
          json['usuarioAtualEhResponsavel'] as bool? ?? false,
      criadoEm: DateTime.parse(json['criadoEm'] as String).toLocal(),
      atualizadoEm: DateTime.parse(json['atualizadoEm'] as String).toLocal(),
    );
  }

  final String identificador;
  final String identificadorDoEncontro;
  final String descricao;
  final String situacao;
  final String identificadorDoUsuarioQueCriou;
  final String? identificadorDoUsuarioResponsavel;
  final String? nomeDoResponsavel;
  final String? urlDaFotoDePerfilDoResponsavel;
  final bool usuarioAtualEhResponsavel;
  final DateTime criadoEm;
  final DateTime atualizadoEm;

  bool get estaResolvido => situacao.toLowerCase() == 'resolvido';
}

File: EntreNosApp/src/ProjetoEncontros.Dominio/Encontros/ItemDoEncontro.cs
using ProjetoEncontros.Dominio.Compartilhado;

namespace ProjetoEncontros.Dominio.Encontros;

public sealed class ItemDoEncontro : Entidade
{
    public const int TamanhoMaximoDaDescricao = 140;

    private ItemDoEncontro()
    {
        Descricao = "Combinado";
    }

    private ItemDoEncontro(
        Guid identificador,
        Guid identificadorDoEncontro,
        string descricao,
        Guid identificadorDoUsuarioQueCriou,
        Guid? identificadorDoUsuarioResponsavel,
        DateTimeOffset criadoEm)
        : base(identificador, criadoEm)
    {
        if (identificadorDoEncontro == Guid.Empty)
        {
            throw new ExcecaoDeDominioException("O identificador do encontro do item não pode ser vazio.");
        }

        if (identificadorDoUsuarioQueCriou == Guid.Empty)
        {
            throw new ExcecaoDeDominioException("O identificador do usuário que criou o item não pode ser vazio.");
        }

        ValideResponsavel(identificadorDoUsuarioResponsavel);

        IdentificadorDoEncontro = identificadorDoEncontro;
        Descricao = NormalizeDescricao(descricao);
        IdentificadorDoUsuarioQueCriou = identificadorDoUsuarioQueCriou;
        IdentificadorDoUsuarioResponsavel = identificadorDoUsuarioResponsavel;
        Situacao = SituacaoDoItemDoEncontro.Pendente;
        AtualizadoEm = criadoEm;
    }

    public Guid IdentificadorDoEncontro { get; private set; }

    public string Descricao { get; private set; }

    public Guid IdentificadorDoUsuarioQueCriou { get; private set; }

    public Guid? IdentificadorDoUsuarioResponsavel { get; private set; }

    public SituacaoDoItemDoEncontro Situacao { get; private set; }

    public DateTimeOffset AtualizadoEm { get; private set; }

    public bool EstaPendente
    {
        get
        {
            return Situacao == SituacaoDoItemDoEncontro.Pendente;
        }
    }

    public bool EstaResolvido
    {
        get
        {
            return Situacao == SituacaoDoItemDoEncontro.Resolvido;
        }
    }

    public static ItemDoEncontro Crie(
        Guid identificador,
        Guid identificadorDoEncontro,
        string descricao,
        Guid identificadorDoUsuarioQueCriou,
        Guid? identificadorDoUsuarioResponsavel,
        DateTimeOffset criadoEm)
    {
        return new(
            identificador,
            identificadorDoEncontro,
            descricao,
            identificadorDoUsuarioQueCriou,
            identificadorDoUsuarioResponsavel,
            criadoEm);
    }

    public void AltereResponsavel(Guid? identificadorDoUsuarioResponsavel, DateTimeOffset atualizadoEm)
    {
        ValideResponsavel(identificadorDoUsuarioResponsavel);

        IdentificadorDoUsuarioResponsavel = identificadorDoUsuarioResponsavel;
        AtualizadoEm = atualizadoEm;
    }

    public void Edite(string descricao, Guid? identificadorDoUsuarioResponsavel, DateTimeOffset atualizadoEm)
    {
        ValideResponsavel(identificadorDoUsuarioResponsavel);

        Descricao = NormalizeDescricao(descricao);
        IdentificadorDoUsuarioResponsavel = identificadorDoUsuarioResponsavel;
        AtualizadoEm = atualizadoEm;
    }

    public void MarqueComoResolvido(DateTimeOffset atualizadoEm)
    {
        Situacao = SituacaoDoItemDoEncontro.Resolvido;
        AtualizadoEm = atualizadoEm;
    }

    public void MarqueComoPendente(DateTimeOffset atualizadoEm)
    {
        Situacao = SituacaoDoItemDoEncontro.Pendente;
        AtualizadoEm = atualizadoEm;
    }

    private static string NormalizeDescricao(string descricao)
    {
        if (string.IsNullOrWhiteSpace(descricao))
        {
            throw new ExcecaoDeDominioException("A descrição do item do encontro é obrigatória.");
        }

        string descricaoNormalizada = descricao.Trim();

        if (descricaoNormalizada.Length > TamanhoMaximoDaDescricao)
        {
            throw new ExcecaoDeDominioException("A descrição do item do encontro não pode ultrapassar 140 caracteres.");
        }

        return descricaoNormalizada;
    }

    private static void ValideResponsavel(Guid? identificadorDoUsuarioResponsavel)
    {
        if (identificadorDoUsuarioResponsavel.HasValue && identificadorDoUsuarioResponsavel.Value == Guid.Empty)
        {
            throw new ExcecaoDeDominioException("O identificador do usuário responsável pelo item não pode ser vazio.");
        }
    }
}
